### ISSUE TITLE
User-provided secret bits are not preserved through resource properties

### ISSUE THREAD

**Issue by t0yv0** [pulumi/pulumi-terraform-bridge#1028]
### What happened?

It appears that values marked secret supplied as inputs may become un-secret in resource outputs, causing the underlying data to be stored in the state un-encrypted. This needs to be fixed.

### Expected Behavior

If a value is sensitive and passed as an input to a bridged provider's resource, it does not appear plain-text in the state file but instead appears encrypted.

### Steps to reproduce

```
YAML_FILE=$(cat <<EOF
name: secret-random-yaml
runtime: yaml
description: A minimal AWS Pulumi YAML program
resources:
  param:
    type: random:RandomPet
    properties:
      length:
        fn::secret: 3
EOF
)

STATE_FILE=$(cat <<EOF
{
    "version": 3,
    "deployment": {
        "manifest": {
            "time": "2023-04-18T14:33:45.340415989-07:00",
            "magic": "2e42a0801cdbcd79fa3ac14fe9f133988f353ea58fe5f5552c838ded91ab09a1",
            "version": "3.63.1-dev.0"
        },
        "secrets_providers": {
            "type": "service",
            "state": {
                "url": "https://api.pulumi.com/",
                "owner": "kyle-pulumi-corp",
                "project": "secret-random-yaml",
                "stack": "dev"
            }
        },
        "resources": [
            {
                "urn": "urn:pulumi:dev::secret-random-yaml::pulumi:pulumi:Stack::secret-random-yaml-dev",
                "custom": false,
                "type": "pulumi:pulumi:Stack",
                "created": "2023-04-18T21:32:09.068277611Z",
                "modified": "2023-04-18T21:32:09.068277611Z"
            },
            {
                "urn": "urn:pulumi:dev::secret-random-yaml::pulumi:providers:random::default",
                "custom": true,
                "id": "11df13d0-eb01-4ade-813c-34ca893cbc0e",
                "type": "pulumi:providers:random",
                "created": "2023-04-18T21:33:45.07783452Z",
                "modified": "2023-04-18T21:33:45.07783452Z"
            },
            {
                "urn": "urn:pulumi:dev::secret-random-yaml::random:index/randomPet:RandomPet::param",
                "custom": true,
                "id": "improved-toucan",
                "type": "random:index/randomPet:RandomPet",
                "inputs": {
                    "length": {
                        "4dabf18193072939515e22adb298388d": "1b47061264138c4ac30d75fd1eb44270",
                        "ciphertext": "AAABAOCkyF+2rhjidEBX1wyjOiZOikdJrR7pZcIs5zM/"
                    }
                },
                "outputs": {
                    "id": "improved-toucan",
                    "length": 2,
                    "separator": "-"
                },
                "parent": "urn:pulumi:dev::secret-random-yaml::pulumi:pulumi:Stack::secret-random-yaml-dev",
                "provider": "urn:pulumi:dev::secret-random-yaml::pulumi:providers:random::default::11df13d0-eb01-4ade-813c-34ca893cbc0e",
                "propertyDependencies": {
                    "length": null
                }
            }
        ]
    }
}
EOF
)
echo "$YAML_FILE" > Pulumi.yaml
echo "$STATE_FILE" > og-secret.json

cat og-secret.json | pulumi stack import && pulumi upYAML_FILE=$(cat <<EOF
name: secret-random-yaml
runtime: yaml
description: A minimal AWS Pulumi YAML program
resources:
  param:
    type: random:RandomPet
    properties:
      length:
        fn::secret: 3
EOF
)

STATE_FILE=$(cat <<EOF
{
    "version": 3,
    "deployment": {
        "manifest": {
            "time": "2023-04-18T14:33:45.340415989-07:00",
            "magic": "2e42a0801cdbcd79fa3ac14fe9f133988f353ea58fe5f5552c838ded91ab09a1",
            "version": "3.63.1-dev.0"
        },
        "secrets_providers": {
            "type": "service",
            "state": {
                "url": "https://api.pulumi.com/",
                "owner": "kyle-pulumi-corp",
                "project": "secret-random-yaml",
                "stack": "dev"
            }
        },
        "resources": [
            {
                "urn": "urn:pulumi:dev::secret-random-yaml::pulumi:pulumi:Stack::secret-random-yaml-dev",
                "custom": false,
                "type": "pulumi:pulumi:Stack",
                "created": "2023-04-18T21:32:09.068277611Z",
                "modified": "2023-04-18T21:32:09.068277611Z"
            },
            {
                "urn": "urn:pulumi:dev::secret-random-yaml::pulumi:providers:random::default",
                "custom": true,
                "id": "11df13d0-eb01-4ade-813c-34ca893cbc0e",
                "type": "pulumi:providers:random",
                "created": "2023-04-18T21:33:45.07783452Z",
                "modified": "2023-04-18T21:33:45.07783452Z"
            },
            {
                "urn": "urn:pulumi:dev::secret-random-yaml::random:index/randomPet:RandomPet::param",
                "custom": true,
                "id": "improved-toucan",
                "type": "random:index/randomPet:RandomPet",
                "inputs": {
                    "length": {
                        "4dabf18193072939515e22adb298388d": "1b47061264138c4ac30d75fd1eb44270",
                        "ciphertext": "AAABAOCkyF+2rhjidEBX1wyjOiZOikdJrR7pZcIs5zM/"
                    }
                },
                "outputs": {
                    "id": "improved-toucan",
                    "length": 2,
                    "separator": "-"
                },
                "parent": "urn:pulumi:dev::secret-random-yaml::pulumi:pulumi:Stack::secret-random-yaml-dev",
                "provider": "urn:pulumi:dev::secret-random-yaml::pulumi:providers:random::default::11df13d0-eb01-4ade-813c-34ca893cbc0e",
                "propertyDependencies": {
                    "length": null
                }
            }
        ]
    }
}
EOF
)
echo "$YAML_FILE" > Pulumi.yaml
echo "$STATE_FILE" > og-secret.json

cat og-secret.json | pulumi stack import && pulumi up
```

### Output of `pulumi about`

N/A

### Additional context

N/A

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by t0yv0** [comment]
Note: besides secrets that originate from the Pulumi program (and the engine), also ensure that properties marked as secret in the schema are appropriately marked as such in the output.

**Comment by iwahbe** [comment]
@t0yv0 Should we think of this as a P1?

**Comment by t0yv0** [comment]
I guess you're right. Possibly. This was courtesy @dixler and @justinvp 's team is discussing engine-level fix. 

Would be good to see if this applies not only to PF but normal providers too (random is PF-based).

**Comment by t0yv0** [comment]
I've checked on pulumi-gcp and it's not affected. This is a plugin-framework scoped bug I think.

**Comment by iwahbe** [comment]
> Note: besides secrets that originate from the Pulumi program (and the engine), also ensure that properties marked as secret in the schema are appropriately marked as such in the output.

SDKs are responsible for communicating input secretness. We can assume that if an input is marked secret in the schema, it will be sent as a secret.

**Comment by iwahbe** [comment]
@t0yv0 Can you sit down with me tomorrow and help me repro? The repro you have described fails for me with this error (I have logged in):
```
error: could not deserialize deployment: constructing secrets manager of type "service": could not find access token for https://api.pulumi.com/, have you logged in?
```

Manually running code such as:
```yaml
name: secret-random-yaml
runtime: yaml
resources:
  param:
    type: random:RandomPet
    properties:
      length:
        fn::secret: 3
```

has the expected state:

```json
{
    "version": 3,
    "deployment": {
        "manifest": {
            "time": "2023-04-25T16:58:38.28745-07:00",
            "magic": "3f5be839cf2876cb7e13c81e0e8b2bc76af777b7631a50aa79e6b1c6ca904424",
            "version": "3.64.1-dev.0"
        },
        "secrets_providers": {
            "type": "service",
            "state": {
                "url": "https://api.pulumi.com",
                "owner": "pulumi-test",
                "project": "image-yaml",
                "stack": "test"
            }
        },
        "resources": [
            {
                "urn": "urn:pulumi:test::image-yaml::pulumi:pulumi:Stack::image-yaml-test",
                "custom": false,
                "type": "pulumi:pulumi:Stack",
                "created": "2023-04-25T23:51:35.63181Z",
                "modified": "2023-04-25T23:51:35.63181Z"
            },
            {
                "urn": "urn:pulumi:test::image-yaml::pulumi:providers:random::default",
                "custom": true,
                "id": "ebcc693b-efa0-4032-9c84-f4b690918fcd",
                "type": "pulumi:providers:random",
                "created": "2023-04-25T23:51:35.978393Z",
                "modified": "2023-04-25T23:51:35.978393Z"
            },
            {
                "urn": "urn:pulumi:test::image-yaml::random:index/randomPet:RandomPet::res",
                "custom": true,
                "id": "jointly-relieved-mackerel",
                "type": "random:index/randomPet:RandomPet",
                "inputs": {
                    "__defaults": [
                        "separator"
                    ],
                    "length": {
                        "4dabf18193072939515e22adb298388d": "1b47061264138c4ac30d75fd1eb44270",
                        "ciphertext": "AAABANL7kfmJg1dakhjT9Qe3+ht8pSjcURMQW+ORAyD4"
                    },
                    "separator": "-"
                },
                "outputs": {
                    "id": "jointly-relieved-mackerel",
                    "length": {
                        "4dabf18193072939515e22adb298388d": "1b47061264138c4ac30d75fd1eb44270",
                        "ciphertext": "AAABAFS52YiCjLKOt1VjODknOsODCRHIzoZORs59VneP"
                    },
                    "separator": "-"
                },
                "parent": "urn:pulumi:test::image-yaml::pulumi:pulumi:Stack::image-yaml-test",
                "provider": "urn:pulumi:test::image-yaml::pulumi:providers:random::default::ebcc693b-efa0-4032-9c84-f4b690918fcd",
                "propertyDependencies": {
                    "length": null
                },
                "created": "2023-04-25T23:58:38.16983Z",
                "modified": "2023-04-25T23:58:38.16983Z"
            }
        ]
    }
}
```

**Comment by t0yv0** [comment]
I have investigated this further. 

The v3 tfbridge does NOT preserve value based secrets in resource-based methods, see:

https://github.com/pulumi/pulumi-terraform-bridge/compare/master...t0yv0/test-secret-preservation

How does this work then? I've investigated this program:

```yaml
name: secret-random-yaml
runtime: yaml
description: A minimal AWS Pulumi YAML program
resources:
  param:
    type: random:RandomPet
    properties:
      length:
        fn::secret: 3
```

under two versions of provider random:

```
NAME    KIND      VERSION  SIZE   INSTALLED       LAST USED
random  resource  4.13.0   72 MB  12 minutes ago  11 minutes ago
random  resource  4.8.1    31 MB  16 minutes ago  15 minutes ago
```

and found this key discrepancy:

```
# 4.8.1
{
  "method": "/pulumirpc.ResourceProvider/Configure",
  "request": {
    "args": {},
    "acceptSecrets": true,
    "acceptResources": true
  },
  "response": {
    "supportsPreview": true
  },
  "metadata": {
    "kind": "resource",
    "mode": "client",
    "name": "random"
  }
}

# 4.13.1
{
  "method": "/pulumirpc.ResourceProvider/Configure",
  "request": {
    "args": {},
    "acceptSecrets": true,
    "acceptResources": true
  },
  "response": {
    "acceptSecrets": true,
    "supportsPreview": true,
    "acceptResources": true
  },
  "metadata": {
    "kind": "resource",
    "mode": "client",
    "name": "random"
  }
}
```

Notice that "acceptSecrets": true that PF-based version states makes secrets appear in the inputs of Create. Apparently using acceptSecrets: false make the engine apply a heuristic to mark the appropriate values as secret. 

## Suggested fix

The right call for us is to fix Plugin Framework branch to use "acceptSecrets": false and leave the heuristics to the engine (or else provider SDKs maintained in pulumi/pulumi), the we do no worse or exactly the same as V3 bridge.

## Further Issues

The responsibility of this repository for now remains to ensure that output paths marked secret in the schema are returned as such to the engine, because the engine does not consult the schema.

- pkg/v3
   - For Provider Config this recently was extended to nested properties https://github.com/pulumi/pulumi-terraform-bridge/issues/974 but remains overly conservative TODO issues
   - For Resources this is handled in by MakeTerraformOutputs supportsSecrets parameter
   - For DataSources this is blocked by https://github.com/pulumi/pulumi/issues/12710

- pf
  - Provider config: TBD
  - Resources: TBD
  - For DataSources this is blocked by https://github.com/pulumi/pulumi/issues/12710

Separately we need to understand what acceptResources means and whether PF should set that to true or false.

## Future work: secrets heuristics

There are several versions of the heursitic proposed and I think it warrants a design discussion to get it right, in the engine and/or plugin SDKs:

- @dixler proposed a heuristic that makes property X secret in a resource output if the resource received a secret input X, is there a reference ticket for the proposal?

- @t0yv0 when working on fix for https://github.com/pulumi/pulumi-terraform-bridge/issues/974 conservatively implemented a value-based heuristic that decided if input property X is secret if it ContainsSecrets anywhere deep down, and conservatively would mark the output X (of the same name) as secret if that is the case. This sounds like a variation of @dixler proposal, just applied o provider-level config. In retrospect this is likely not necessary (checking up on this) since bridged providers do not accept secrets and the engine should be doing the heuristic for secrets in this case.

- @justinvp proposed a heuristic that scans for secret values in inputs and then marks any similar values in outputs as secret. Is there a reference ticket? This would be my preferred proposal if we can extend it to scan for secret object  keys and mark any object/map as a secret if it contains any secret keys. I'm happy to write a pulumi/pulumi ticket for this if there isn't one.

- @iwahbe proposed an implemented a path-based heuristic that scans inputs for paths where secrets are found and marks these paths as secret in the outputs; failing to find the path in the output the heuristic recurs into a path prefix to "broaden the blast radius" of what is marked as secret; this is implemented in https://github.com/pulumi/pulumi-terraform-bridge/pull/1035 - we had a brief ideation on this but I have second thoughts on the applicability of widening

**Comment by t0yv0** [comment]
Filed https://github.com/pulumi/pulumi-terraform-bridge/issues/1051 - this is the most significant user-visible limitation of current status quo.

**Comment by t0yv0** [comment]
https://github.com/pulumi/home/issues/2812 for chasing down the design questions later.
